Unicorn stream8 feature bjs2 52586

alobraztsov86 opened the pull request.

Commit:
initial commit after creating project at local PC

Commit:
BJS-52586 Recommendations Добавление рекомендации. Exception autotest works.

Commit:
Merge remote-tracking branch 'origin/unicorn-master-stream8' into unicorn-master-stream8

TFdevQA commented on src/main/java/school/faang/user_service/controller/RecommendationController.java:
```diff
@@ -0,0 +1,25 @@
+
+    public RecommendationDto giveRecommendation(RecommendationDto recommendation) {
+        if (recommendationDtoIsValid(recommendation)) {
+            return recommendationService.create(recommendation);// Как вызывать? Solved   //? как-то вернуть  RecommentationDto object without using 'new'.
```
Надо смапить dto в  entity и передать в service

alobraztsov86 replied:
@TFdevQA  Тут мне пока не понятен замысел такого действия.

TFdevQA commented on src/main/java/school/faang/user_service/mapper/RecommendationMapper.java:
```diff
@@ -0,0 +1,24 @@
+//    private User getAuthor(User author) {
+//        return userRepository.findAll(author.getId());
+//    }
+}
```
комментарии надо удалять
в жизни автоматические чекеры не пропускают код с комментами

alobraztsov86 replied:
Комментарии я пока оставляю, тк без них через короткое время я сам с трудом понимаю, что я там делаю, и к какому пункту требований относится код

TFdevQA commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,99 @@
+        // автор дает рекомендацию не раньше, чем через 6 месяцев после его последней рекомендации этому пользователю.
+        // Также проверить, что навыки, предлагаемые в рекомендации существуют в системе.
+        Long newRecommendationId;
+        if (LocalDateTime.now().minusMonths(6).isAfter(recommendation.getCreatedAt())
```
а как ты здесь проверяешь инфу из дто (то есть то, что пришло в реквесте)? Надо же из базы достать сначала record и у нее смотреть последнюю дату рекомендации

alobraztsov86 replied:
@TFdevQA Спасибо, сделаю!
Хоть это и четко не прописано в требованиях, но очень важный функционал имеет смысл.

TFdevQA commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,99 @@
+        userSkillGuaranteeRepository.create(recommendationDto.getId(),
+                skillOfferDto.getSkillId(),
+                recommendationDto.getAuthorId());
+    }
```
А где обновление даты последней рекомендации?

alobraztsov86 replied:
В рамках требований  к Разделу "Дать рекомендацию" - 
обновление даты последней рекомендации не прописано. Я не нашёл как - минимум.

Раздел "Обновить рекомендацию" буду делать в следующем Pull Request

alobraztsov86 replied:
Написал метод update() для обновления рекомендации

alobraztsov86 replied:
Оказывается, время должно автоматически обновиться, тк есть у поля updatedAt аннотация  @UpdateTimestamp в классе Recommendation

TFdevQA commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,99 @@
+                    recommendation.getReceiverId(),
+                    recommendation.getContent(),
+                    recommendation.getSkillOffers(),
+                    recommendation.getCreatedAt());
```
что-то совсем запутано- у тебя recommendation это и есть дто, которое ты принял
И если здесь подразумевалось преобразование из entity to dto, то тогда для этого и нужен mapping

alobraztsov86 replied:
Тут, единственный смысл, который вкладывал в такую запись - Добавить значение **id**, полученное после создания объекта в базе данных. А маппер не написал, т.к. не видел в нём смысла и думал, что у нас мапперы только для преобразования из Dto в Entity и наоборот. А тут нет такой необходимости. 

Но можно получить Entity и смапить его на Dto.

**Сделал**
```
Optional<Recommendation> recommendationFromDataBase = recommendationRepository.findById(newRecommendationId);
            return recommendationFromDataBase.map(recommendationMapper::toDto).orElse(null);
```

TFdevQA reviewed (changes requested):
У тебя конфликт - убери из комита файл UserServiceApplication

Commit:
BJS-52586 Recommendations. методы updateRecommendation(), deleteRecommendation(), getAllUserRecommendations(), getAllGivenRecommendations(),

Commit:
BJS-52586 Recommendations. методы addGuarantorToSkill()
addSkillAndAddGuarantor()

Commit:
BJS-52586 Recommendations. Simplified delete() and update() methods.
Autotests 100% for RecommendationController.
DTO goes to RecommendationService class.

Commit:
BJS-52586 Recommendations. DTO to Entity in RecommendationController class.
Reworked autotests for RecommendationController

Commit:
BJS-52586 Recommendations. Autotests for RecommendationService
Refactored method RecommendationService.create() and update().

Commit:
BJS-52586 Recommendations. unused imports are deleted

alobraztsov86 commented:
> У тебя конфликт - убери из комита файл UserServiceApplication

Вроде он перетёрся при мёрже мастера в мою ветку. Но это не точно.

TFdevQA commented on src/main/java/school/faang/user_service/dto/recommendation/RecommendationDto.java:
```diff
@@ -0,0 +1,27 @@
+        this.content = "";
+        this.skillOffers = null;
+        this.createdAt  = null;
+    }
```
Зачем эта инициализация всем null ?

Nafanuil replied:
у меня тот же вопрос

alobraztsov86 replied:
Была ошибка как-то, что spring bean не мог создать объект
RecommendationDto из-за отсутствия конструктора без параметров, а на аннотацию @NoArgsConstructor
ругался компилятор. 

Сейчас поставил  @NoArgsConstructor и работает.

Nafanuil commented on src/main/java/school/faang/user_service/UserServiceApplication.java:
```diff
@@ -14,7 +14,7 @@ public class UserServiceApplication {
         SpringApplication.run(UserServiceApplication.class, args);
     }
-
+    
```
что то лишнее)

Nafanuil replied:
на будущее старайся что бы в твой мр не попадало ничего лишнего, такие файлы только мешают +строка - строчка, это вопрос форматирования и он должен быть настроен у всех одинаково

Nafanuil commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationController.java:
```diff
@@ -0,0 +1,87 @@
+import java.util.List;
+
+@Controller
+//@RequiredArgsConstructor
```
что то лишнее)

Nafanuil commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationController.java:
```diff
@@ -0,0 +1,87 @@
+@AllArgsConstructor
+public class RecommendationController {
+    @Autowired
+    RecommendationService recommendationService;
```
а чего он не private final  ?

alobraztsov86 replied:
исправил на private  final

Nafanuil commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationController.java:
```diff
@@ -0,0 +1,87 @@
+
+    @PostMapping("/giverecommendation")
+    @ResponseBody
+    public RecommendationDto giveRecommendation(@RequestBody RecommendationDto recommendationDto) {
```
почему то весь мапинг из ентити в дто остался в сервисе, давай вернем его на свое место(контроллер) везде

alobraztsov86 replied:
Сделал для create() and update() методов

Nafanuil replied:
а для getAllUserRecommendations и getAllGivenRecommendations чего не сделал?

Nafanuil commented on src/main/java/school/faang/user_service/mapper/RecommendationMapper.java:
```diff
@@ -0,0 +1,20 @@
+import school.faang.user_service.dto.recommendation.RecommendationDto;
+import school.faang.user_service.entity.recommendation.Recommendation;
+
+@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE, uses=SkillOfferMapper.class)
```
поменяй на ReportingPolicy.ERROR

alobraztsov86 replied:
поменял, спасибо

Nafanuil commented on src/main/java/school/faang/user_service/mapper/SkillOfferMapper.java:
```diff
@@ -0,0 +1,17 @@
+import school.faang.user_service.dto.recommendation.SkillOfferDto;
+import school.faang.user_service.entity.recommendation.SkillOffer;
+
+@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE, uses = SkillMapper.class)
```
поменяй на ReportingPolicy.ERROR

Nafanuil commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,212 @@
+
+    public RecommendationDto create(Recommendation recommendation) {
+
+        Long newRecommendationId;
```
а в чем смысл отделение объявления переменной от ее присваивания?

Nafanuil commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,212 @@
+                recommendation.getAuthor().getId(),
+                recommendation.getReceiver().getId());
+        LocalDateTime timeOfLastRecommendationGivenToReceiver = recommendationFromDB.map(Recommendation::getUpdatedAt)
+                .orElse(LocalDateTime.of(0, 1, 1, 0, 0));
```
если посмотреть как это поле объявлено в базе то можно увидеть что если его туда не положить то будет браться дефолтное значение (но это просто для расширения кругозора), 

мой вопрос в том что я никак не могу понять что за дата такая LocalDateTime.of(0, 1, 1, 0, 0) ?

alobraztsov86 replied:
Спасибо! посмотрю какое дефолтное.
LocalDateTime.of(0, 1, 1, 0, 0) - дата которая точно очень давняя, чтобы метод возвращал true, чтобы не блокировать добавление/изменение рекомендацци

Nafanuil commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,212 @@
+
+        restrictionTimePeriodForNewRecommendationIsOver(recommendation);
+        skillsArePresentedInSystem(recommendation.getSkillOffers());
+        recommendationRepository.updateByRecommendationId(
```
хмммм, то есть  получается что обновление сущъности в бд идет двумя запросами, я сейчас не говорю что нельзя писать свои запросы, но раз ты его уже написал почему бы не сделать его цельным, что бы не надо было два раза ходить в дб и обновлять одну и туже строчку, ну или починить метод save)))

alobraztsov86 replied:
Тут я ,похоже забыл убрать вызов одного из 2х методов)) 
Перемещу вызов 
recommendationRepository.updateByRecommendationId() 
в
updateRecommendation(recommendation)

Nafanuil commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,212 @@
+     */
+    @Transactional
+    public void delete(long id) {
+        Optional<Recommendation> optional = recommendationRepository.findById(id);
```
а зачем тут лишний поход в базу? 

пытаюсь придумать кейс где бы мне надо было как то обработать ситуацию что я пытаюсь удалить запись которой нет и мне что то надо с этим сделать но не могу

Nafanuil commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,212 @@
+    }
+
+    public List<RecommendationDto> getAllUserRecommendations(long receiverId) {
+        Page<Recommendation> recommendationPage = recommendationRepository.findAllByReceiverId(
```
в коде который мы вам дали есть явная не состыковка с тем что нужно сделать (я написал по этому поводу что бы исправили)

давай пока добавим в репозиторий метод List<Recommendation> findAllByReceiverId(long receiverId);
и будет использовать его

иначе получается что мы взяли только 1 страничку а не все рекомендации, ведь их может быть больше чем помещается на 1 страничку

Плюсик тебе большой за то что разобрался как пагинацией работать

alobraztsov86 replied:
Переделаю, спасибо.

Nafanuil commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,212 @@
+    }
+
+    public List<RecommendationDto> getAllGivenRecommendations(long authorId) {
+        Page<Recommendation> recommendationPage = recommendationRepository.findAllByAuthorId(
```
с этим методом предлагаю сделать тоже самое что и с верхним

Nafanuil commented on src/main/java/school/faang/user_service/mapper/SkillOfferMapper.java:
```diff
@@ -0,0 +1,17 @@
+import school.faang.user_service.entity.recommendation.SkillOffer;
+
+@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE, uses = SkillMapper.class)
+public interface SkillOfferMapper {
```
посмотри на https://mapstruct.org/documentation/stable/reference/html/#mappings-with-several-source-parameters это должно навести на мысль как смапить лучше

alobraztsov86 replied:
Добавил
@Mapping(source = "skill.title", target = "title")
@Mapping(source = "recommendation.receiver.id", target = "receiverUserId")
@Mapping(source = "recommendation.author.id", target = "requesterUserId")

Наверно, так лучше.

alobraztsov86 replied:
Из документации не очень понял, как можно улучшить, но может этими изменениями попал и сделал лучше, как минимум сервер теперь присылает обратно полностью заполненные SkillOffers

Nafanuil commented on src/main/java/school/faang/user_service/mapper/SkillOfferMapper.java:
```diff
@@ -0,0 +1,17 @@
+    @Mapping(source = "skill.id", target = "skillId")
+    SkillOfferDto toDto(SkillOffer skillOffer);
+
+    @Mapping(source = "skillId", target = "skill.id")
```
вот тут концептуально немного не правильно, ты используешь SkillMapper для конструирования Skill а потом домапливаешь руками еще 2 поля

там есть прикол у мапстракта что если ты в target что то мапишь через "." то он поднимает лапки и больше не проверяет замапил ли ты остальные поля у этого поля или нет, а это значит ReportingPolicy.ERROR больше не работает как надо, собственно поэтому лучше избигать мапить что то через "." а вместо этого по максимуму использовать другие маперы

alobraztsov86 replied:
Здесь убрал uses = SkillMapper.class из аннтации @Mapper
 и добавил
@Mapping(target = "recommendation", ignore = true)

Nafanuil replied:
можно сделать вот такую вот штуку 


    @Mapping(source = "skillOfferDto.skillId", target = "skill.id")
    @Mapping(source = "skillOfferDto.title", target = "skill.title")
    @Mapping(source = "skillOfferDto.id", target = "id")
    SkillOffer toEntity(SkillOfferDto skillOfferDto, Recommendation recommendation);

    List<SkillOffer> toEntity(List<SkillOfferDto> skillOfferDto, @Context Recommendation recommendation);

    default SkillOffer toEntityWithContext(SkillOfferDto skillOfferDto, @Context Recommendation recommendation) {
        return toEntity(skillOfferDto, recommendation);
    }


а в RecommendationMapper сделать вот так 

    @Mapping(target = "skillOffers", expression = "java(skillOfferMapper.toEntity(recommendationDto.getSkillOffers(), recommendation))")

Nafanuil commented on src/main/java/school/faang/user_service/repository/SkillRepository.java:
```diff
@@ -3,11 +3,12 @@
 import java.util.List;
 import java.util.Optional;
-
+@Transactional
```
тут эта аннотация не нужна

Nafanuil commented on src/main/java/school/faang/user_service/repository/UserSkillGuaranteeRepository.java:
```diff
@@ -1,7 +1,33 @@
+            DELETE FROM user_skill_guarantee u
+            WHERE u.user_id = ?1 AND u.skill_id = ?2 AND u.guarantor_id=?3
+            """)
+    void deleteSkillGuaranteedToUserByGuarantee(long userId, long skillId, long guarantorId);
```
не найду где этот метод вызывается

Nafanuil commented on src/main/java/school/faang/user_service/repository/UserSkillGuaranteeRepository.java:
```diff
@@ -1,7 +1,33 @@
+            INSERT INTO user_skill_guarantee (user_id, skill_id, guarantor_id)
+            VALUES (?1, ?2, ?3) returning id
+            """)
+    Long create(long userId, long skillId,long guarantorId);
```
у тебя тут уже есть такой метод, save , делает ровно тоже самое

Nafanuil commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,212 @@
+
+        newRecommendationId = saveRecommendation(recommendation);
+        saveSkillOffers(recommendation.getSkillOffers(), newRecommendationId);
+        addSkillsAndAddGuarantor(recommendation, newRecommendationId);
```
вот этот метод у меня вызывает какие то странные ощущения, давай пока поправим все остальное что я понаписал тут а к нему вернемся потом, есть вероятность что он сам сабой удалится к этому времени

alobraztsov86 replied:
Хорошо, может метод save()  и аннотации Hybernate могут сами всё сделать, но пока ещё не делают.

Commit:
BJS-52586 Recommendations. Review updates

Commit:
BJS-52586 Recommendations. Autotests fixed. Service.delete() method updated

Commit:
BJS-52586 Recommendations. unused imports deleted

Nafanuil commented on src/main/java/school/faang/user_service/repository/SkillRepository.java:
```diff
@@ -3,6 +3,7 @@
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.data.jpa.repository.Modifying;
 import org.springframework.data.jpa.repository.Query;
+import org.springframework.transaction.annotation.Transactional;
```
лишний код

Nafanuil commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,206 @@
+    private final UserSkillGuaranteeRepository userSkillGuaranteeRepository;
+    @Autowired
+    private final RecommendationMapper recommendationMapper;
+    private static final int PAGE_NUMBER = 0;
```
Page мы больше не используем а не используемый код остался
